src/lib/Handshake.js

```diff
@@ -87,76 +88,81 @@ const ID_BYTES = 12;
 const SESSION_BYTES = 40;
 const GLOBAL_TIMEOUT = 10;
 
+// TODO make Handshake module stateless.
 class Handshake {
   _client: Device;
+  _deviceKeyRepository: Repository<string>;
   _socket: Socket;
   _handshakeStage: HandshakeStage = 'send-nonce';
   _reject: ?Function;
-  _deviceID: string = '';
+  _deviceID: string;
   _pendingBuffers: Array<Buffer> = [];
   _useChunkingStream: boolean = true;
 
-  constructor(client: Device) {
-    this._client = client;
-    this._socket = client._socket;
+  constructor(deviceKeyRepository: Repository<string>) {
+    this._deviceKeyRepository = deviceKeyRepository;
   }
 
-  start = async (): Promise<*> => {
+  start = async (device: Device): Promise<*> => {
+    this._client = device;
+    this._socket = device._socket;
+
     return Promise.race([
       this._runHandshake(),
       this._startGlobalTimeout(),
       new Promise((resolve, reject) => this._reject = reject),
-    ]).catch((message) => {
+    ]).catch((error) => {
       var logInfo = {
         cache_key: this._client && this._client._connectionKey,
         ip: this._socket && this._socket.remoteAddress
           ? this._socket.remoteAddress.toString()
           : 'unknown',
-        deviceID: this._deviceID ? this._deviceID.toString('hex') : null,
+        deviceID: this._deviceID || null,
       };
 
-      logger.error('Handshake failed: ', message, logInfo);
+      logger.error('Handshake failed: ', error, logInfo);
 
-      throw message;
+      throw error;
     });
   };
 
   _runHandshake = async (): Promise<*> => {
-    try {
-      const dataAwaitable = this._onSocketDataAvailable();
       const nonce = await this._sendNonce();
-      const data = await dataAwaitable;
-      const deviceProvidedPem = this._readDeviceID(nonce, data);
-      const publicKey = this._getDeviceKey(nullthrows(deviceProvidedPem));
+      const data = await this._onSocketDataAvailable();
+
+      const {
+        deviceID,
+        deviceProvidedPem,
+      } = this._readDeviceHandshakeData(nonce, data);
+      this._deviceID = deviceID;
```

Comment:
also...i do this now only for log:
https://github.com/Brewskey/spark-protocol/pull/25/files#diff-cb4067f4badfa0c989e0cf0856a9a8baR115